File: update-core/src/main/java/dev/bee/kanjianki/updatecore/UpdateReleaseAssetSelector.java
package dev.bee.kanjianki.updatecore;

import java.util.Objects;

public final class UpdateReleaseAssetSelector {
    private UpdateReleaseAssetSelector() {
    }

    public static AssetSelection selectAssets(GitHubReleaseMetadata release) {
        if (release == null) {
            return AssetSelection.failure("Latest release metadata is empty.");
        }
        GitHubReleaseMetadata.ReleaseAsset apk = apkAsset(release);
        if (apk == null) {
            return AssetSelection.failure("Latest release has no APK asset.");
        }
        GitHubReleaseMetadata.ReleaseAsset checksum = checksumAssetFor(release, apk.name());
        if (checksum == null) {
            return AssetSelection.failure("Latest release has no SHA-256 checksum asset.");
        }
        return AssetSelection.success(apk, checksum);
    }

    private static GitHubReleaseMetadata.ReleaseAsset apkAsset(GitHubReleaseMetadata release) {
        for (GitHubReleaseMetadata.ReleaseAsset asset : release.assets()) {
            if (asset.name().endsWith(".apk")) {
                return asset;
            }
        }
        return null;
    }

    private static GitHubReleaseMetadata.ReleaseAsset checksumAssetFor(GitHubReleaseMetadata release, String apkName) {
        for (GitHubReleaseMetadata.ReleaseAsset asset : release.assets()) {
            if (Objects.equals(asset.name(), apkName + ".sha256")) {
                return asset;
            }
        }
        return null;
    }

    public static final class AssetSelection {
        private final boolean ok;
        private final String message;
        private final GitHubReleaseMetadata.ReleaseAsset apk;
        private final GitHubReleaseMetadata.ReleaseAsset checksum;

        private AssetSelection(
                boolean ok,
                String message,
                GitHubReleaseMetadata.ReleaseAsset apk,
                GitHubReleaseMetadata.ReleaseAsset checksum
        ) {
            this.ok = ok;
            this.message = message;
            this.apk = apk;
            this.checksum = checksum;
        }

        public boolean ok() {
            return ok;
        }

        public String message() {
            return message;
        }

        public GitHubReleaseMetadata.ReleaseAsset apk() {
            return apk;
        }

        public GitHubReleaseMetadata.ReleaseAsset checksum() {
            return checksum;
        }

        private static AssetSelection success(
                GitHubReleaseMetadata.ReleaseAsset apk,
                GitHubReleaseMetadata.ReleaseAsset checksum
        ) {
            return new AssetSelection(true, "", apk, checksum);
        }

        private static AssetSelection failure(String message) {
            return new AssetSelection(false, message, null, null);
        }
    }
}

File: update-core/src/main/kotlin/dev/bee/kanjianki/updatecore/UpdateReleaseAssetSelector.kt
package dev.bee.kanjianki.updatecore

import java.util.Locale

object UpdateReleaseAssetSelector {
    private const val JAPANESE_LANGUAGE = "ja"

    const val EMPTY_METADATA_MESSAGE = "Latest release metadata is empty."
    const val MISSING_APK_MESSAGE = "Latest release has no APK asset."
    const val MISSING_CHECKSUM_MESSAGE = "Latest release has no SHA-256 checksum asset."

    @JvmStatic
    fun selectAssets(release: GitHubReleaseMetadata?): AssetSelection {
        if (release == null) {
            return AssetSelection.failure(emptyMetadataMessage())
        }
        val apk = apkAsset(release)
            ?: return AssetSelection.failure(missingApkMessage())
        val checksum = checksumAssetFor(release, apk.name())
            ?: return AssetSelection.failure(missingChecksumMessage())
        return AssetSelection.success(apk, checksum)
    }

    @JvmStatic
    fun emptyMetadataMessage(): String = localizedText(
        EMPTY_METADATA_MESSAGE,
        "最新リリースのメタデータが空です。",
    )

    @JvmStatic
    fun missingApkMessage(): String = localizedText(
        MISSING_APK_MESSAGE,
        "最新リリースにAPKアセットがありません。",
    )

    @JvmStatic
    fun missingChecksumMessage(): String = localizedText(
        MISSING_CHECKSUM_MESSAGE,
        "最新リリースにSHA-256チェックサムのアセットがありません。",
    )

    private fun apkAsset(release: GitHubReleaseMetadata): GitHubReleaseMetadata.ReleaseAsset? {
        return release.assets().firstOrNull { asset -> asset.name().endsWith(".apk") }
    }

    private fun checksumAssetFor(
        release: GitHubReleaseMetadata,
        apkName: String,
    ): GitHubReleaseMetadata.ReleaseAsset? {
        return release.assets().firstOrNull { asset -> asset.name() == "$apkName.sha256" }
    }

    private fun localizedText(english: String, japanese: String): String =
        if (Locale.getDefault().language == JAPANESE_LANGUAGE) japanese else english

    class AssetSelection private constructor(
        @JvmField val ok: Boolean,
        @JvmField val message: String,
        @JvmField val apk: GitHubReleaseMetadata.ReleaseAsset?,
        @JvmField val checksum: GitHubReleaseMetadata.ReleaseAsset?,
    ) {
        fun ok(): Boolean = ok
        fun message(): String = message
        fun apk(): GitHubReleaseMetadata.ReleaseAsset = requireNotNull(apk)
        fun checksum(): GitHubReleaseMetadata.ReleaseAsset = requireNotNull(checksum)

        companion object {
            fun success(
                apk: GitHubReleaseMetadata.ReleaseAsset,
                checksum: GitHubReleaseMetadata.ReleaseAsset,
            ): AssetSelection {
                return AssetSelection(true, "", apk, checksum)
            }

            fun failure(message: String): AssetSelection {
                return AssetSelection(false, message, null, null)
            }
        }
    }
}
